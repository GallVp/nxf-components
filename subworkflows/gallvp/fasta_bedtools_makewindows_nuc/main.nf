include { SAMTOOLS_FAIDX            } from '../../../modules/gallvp/samtools/faidx/main'
include { BEDTOOLS_MAKEWINDOWS      } from '../../../modules/gallvp/bedtools/makewindows/main'
include { BEDTOOLS_NUC              } from '../../../modules/gallvp/bedtools/nuc/main'

workflow FASTA_BEDTOOLS_MAKEWINDOWS_NUC {

    take:
    ch_fasta                        // channel: [ val(meta), fasta ]

    main:

    ch_versions = Channel.empty()

    // MODULES: SAMTOOLS_FAIDX
    SAMTOOLS_FAIDX (
        ch_fasta,
        [ [], [] ],
        true // get_sizes
    )

    ch_sizes                        = SAMTOOLS_FAIDX.out.sizes
    ch_versions                     = ch_versions.mix(SAMTOOLS_FAIDX.out.versions.first())

    // collectFile: Regions BED
    ch_regions_bed                  = ch_sizes
                                    | map { meta, sizes ->
                                        [
                                            "${meta.id}.regions.bed",
                                            sizes.readLines().collect { row ->
                                                def cols = row.tokenize('\t')
                                                def name = cols[0]
                                                def size = cols[1]
                                                def start = 0
                                                def end = size // BED is 0-indexed end-exclusive

                                                "${name}\t${start}\t${end}"
                                            }.join('\n')
                                        ]
                                    }
                                    | collectFile(newLine: true)
                                    | map { file ->
                                        [
                                            [ id: file.baseName.replace('.regions', '') ], // meta2: [ id: meta.id ]
                                            file
                                        ]
                                    }

    // MODULE: BEDTOOLS_MAKEWINDOWS
    BEDTOOLS_MAKEWINDOWS (
        ch_regions_bed
    )

    ch_intervals_bed                = BEDTOOLS_MAKEWINDOWS.out.bed
    ch_versions                     = ch_versions.mix(BEDTOOLS_MAKEWINDOWS.out.versions.first())

    // MODULE: BEDTOOLS_NUC
    ch_bedtools_nuc_inputs          = ch_fasta
                                    | map { meta, fasta -> [ [ id: meta.id ], fasta ] }
                                    | join(
                                        ch_intervals_bed
                                    )
    BEDTOOLS_NUC(
        ch_bedtools_nuc_inputs
    )


    ch_versions                     = ch_versions.mix(BEDTOOLS_NUC.out.versions.first())

    emit:
    nuc                             = BEDTOOLS_NUC.out.bed  // channel: [ val(meta2), bed ]
    versions                        = ch_versions           // channel: [ versions.yml ]
}
