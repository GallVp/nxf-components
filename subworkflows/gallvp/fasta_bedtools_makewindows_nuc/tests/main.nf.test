nextflow_workflow {

    name "Test Subworkflow FASTA_BEDTOOLS_MAKEWINDOWS_NUC"
    script "../main.nf"
    config "./nextflow.config"
    workflow "FASTA_BEDTOOLS_MAKEWINDOWS_NUC"

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/fasta_bedtools_makewindows_nuc"
    tag "samtools"
    tag "samtools/faidx"
    tag "bedtools"
    tag "bedtools/makewindows"
    tag "bedtools/nuc"


    test("sarscov2 - fasta") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("sarscov2 - fasta - stub") {
        options '-stub'

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
