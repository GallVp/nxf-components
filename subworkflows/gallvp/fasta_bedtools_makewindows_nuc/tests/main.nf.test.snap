{
    "sarscov2 - fasta - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,17a322435577320c25f6564444a092ec",
                    "versions.yml:md5,20b8080748522e64cd8f91fd7e308d04",
                    "versions.yml:md5,d2ee1f0baf2565c82d1b0687635e19d9"
                ],
                "nuc": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,17a322435577320c25f6564444a092ec",
                    "versions.yml:md5,20b8080748522e64cd8f91fd7e308d04",
                    "versions.yml:md5,d2ee1f0baf2565c82d1b0687635e19d9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-19T11:48:13.426789"
    },
    "sarscov2 - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,977c49e01f2c64deb42bb32398b0c3c9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,20b8080748522e64cd8f91fd7e308d04",
                    "versions.yml:md5,d2ee1f0baf2565c82d1b0687635e19d9",
                    "versions.yml:md5,e2f581f6e9fc8b7a37ada7bd10350228"
                ],
                "nuc": [
                    [
                        {
                            "id": "test"
                        },
                        "test.bed:md5,977c49e01f2c64deb42bb32398b0c3c9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,20b8080748522e64cd8f91fd7e308d04",
                    "versions.yml:md5,d2ee1f0baf2565c82d1b0687635e19d9",
                    "versions.yml:md5,e2f581f6e9fc8b7a37ada7bd10350228"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-19T11:47:23.412481"
    }
}